proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1472 @@
+  
+    storeKey = validator.address
+    storeValue = {
+        "name": validator.name,
+        "totalVotesReceived": totalVotesReceived,
+        "selfVotes": selfVotes,
+        "lastGeneratedHeight": validator.lastGeneratedHeight,
+        "isBanned": validator.isBanned,
+        "pomHeights": validator.pomHeight,
+        "consecutiveMissedBlocks": validator.consecutiveMissedBlocks
+    } serialized using delegateStoreSchema.
+    ```
+    Further, for every entry `validator` in `genesisBlockAssetObject.validators`, also create an entry in the name substore with 
+    ```python
+    storeKey = validator.name utf8-encoded
+    storeValue = {
+        "delegateAddress": validator.address
+    } serialized using nameStoreSchema.
+    ```
+
+* For each entry `voter` in `genesisBlockAssetObject.voters`, create an entry in the voter substore  with
+    ```python
+    storeKey = uint32be(voter.address)
+    storeValue = {
+        "sentVotes": voter.sentVotes,
+        "pendingUnlocks": voter.pendingUnlocks
+    } serialized using voterStoreSchema.
+    ```
+
+* For each entry `snapshot` in `genesisBlockAssetObject.snapshots`, create an entry in the snapshot substore  with
+    ```python
+    storeKey = uint32be(snapshot.roundNumber)
+    storeValue = {
+        "activeDelegates": snapshot.activeDelegates,
+        "delegateWeightSnapshot": snapshot.delegateWeightSnapshot
+    } serialized using snapshotStoreSchema.
+    ```
+
+* Create an entry in the genesis data substore with
+    ```python
+    storeKey = EMPTY_BYTES
+    storeValue = {
+        "height": block header height of the genesis block,
+        "initRounds": genesisBlockAssetObject.GenesisData.initRounds,
+        "initDelegates": genesisBlockAssetObject.GenesisData.initDelegates
+    } serialized using availableLocalIDStoreSchema.
+    ```
+
+#### Genesis State Finalization
+
+To finalize the state of the genesis block the following logic is executed.  If any step fails, the block is discarded and has no further effect.
+
+As in the previous point, let `genesisBlockAssetBytes` be the `data` bytes included in the block assets for the DPoS module and let `genesisBlockAssetObject` be the deserialization of `genesisBlockAssetBytes` according to the `genesisDPoSStoreSchema` schema, given above. Then:
+
+```python
+# register all validators in the Validators module
+for validator in genesisBlockAssetObject.validators:
+    registerValidatorKeys(
+        validator.address,
+        validator.proofOfPossession,
+        validator.generatorKey,
+        validator.blsKey
+    )
+
+# check that all sentVotes and pendingUnlocks correspond to locked tokens
+for address a key of the voter substore:
+    votedAmount = 0
+    for sentVote in voter(address).sentVotes:
+        votedAmount += sentVote.amount
+    for pendingUnlock in voter(address).pendingUnlocks:
+        votedAmount += pendingUnlock.amount
+
+    if getLockedAmount(address, MODULE_ID_DPOS, TOKEN_ID_DPOS) < votedAmount:
```

Comment:
mh, shouldn't we check for an exact match?

Reply:
I would not. This is not a protocol issue and leaves more freedom to developers.

Reply:
maybe I'm missing something, but what kind of freedom are you thinking of? If they want to lock more tokens under the DPoS module they would need to modify the DPoS module, since AFAIK the only locking reason in DPoS is for votes. I think we should be tight when possible

Reply:
Maybe @IkerAlus can break the tie here. I would rather not be tight unless needed.

Reply:
If I was me I would specify it with equality. mainly because I do not see any use case for this.  But I do not have strong opinion, maybe it would be something interesting so we can leave it as it is.

Reply:
Changed to check for equality.